src/storage/src/hummock/store/mod.rs

```diff
@@ -0,0 +1,107 @@
+pub trait FlushFutureTrait<'a> = Future<Output = StorageResult<usize>> + Send;
+pub trait SyncFutureTrait<'a> =
+    Future<Output = StorageResult<(usize, Vec<LocalSstableInfo>)>> + Send;
+
+#[macro_export]
+macro_rules! define_local_state_store_associated_type {
+    () => {
+        type GetFuture<'a> = impl GetFutureTrait<'a>;
+        type IterFuture<'a, R, B>  = impl IterFutureTrait<'a, Self::Iter, R, B>
+                                                            where
+                                                                R: 'static + Send + RangeBounds<B>,
+                                                                B: 'static + Send + AsRef<[u8]>;
+        type InsertFuture<'a> = impl EmptyFutureTrait<'a>;
+        type DeleteFuture<'a> = impl EmptyFutureTrait<'a>;
+        type FlushFuture<'a> = impl FlushFutureTrait<'a>;
+    };
+}
+
+/// State store v2.
+/// It provides the basic functionalities streaming/batch executor needs to access the underlying
+/// state store.
+pub trait StateStore: Send + Sync + 'static + Clone {
+    type Iter: StateStoreIter<Item = (Bytes, Bytes)>;
+
+    type GetFuture<'a>: GetFutureTrait<'a>;
+
+    type IterFuture<'a, R, B>: IterFutureTrait<'a, Self::Iter, R, B>
+    where
+        R: 'static + Send + RangeBounds<B>,
+        B: 'static + Send + AsRef<[u8]>;
+
+    type InsertFuture<'a>: EmptyFutureTrait<'a>;
+
+    type DeleteFuture<'a>: EmptyFutureTrait<'a>;
+
+    type FlushFuture<'a>: FlushFutureTrait<'a>;
+
+    /// Point gets a value from the state store.
+    /// The result is based on a snapshot corresponding to the given `epoch`.
+    fn get(&self, key: &[u8], epoch: u64, read_options: ReadOptions) -> Self::GetFuture<'_>;
+
+    /// Opens and returns an iterator for a given `key_range`.
+    /// The returned iterator will iterate data based on a snapshot corresponding to
+    /// the given `epoch`.
+    fn iter<R, B>(
+        &self,
+        key_range: R,
+        epoch: u64,
+        read_options: ReadOptions,
+    ) -> Self::IterFuture<'_, R, B>
+    where
+        R: RangeBounds<B> + Send,
+        B: AsRef<[u8]> + Send;
+
+    /// Inserts a key-value entry associated with a given `epoch` into the state store.
+    fn insert(&self, key: Bytes, val: Bytes, epoch: u64) -> Self::InsertFuture<'_>;
+
+    /// Deletes a key-value entry from the state store. Only the key-value entry with epoch smaller
+    /// than the given `epoch` will be deleted.
+    fn delete(&self, key: Bytes, epoch: u64) -> Self::DeleteFuture<'_>;
+
+    /// Triggers a flush to persistent storage for the in-memory states.
+    fn flush(&self) -> Self::FlushFuture<'_>;
+}
+
+#[allow(unused)]
+#[derive(Default, Clone)]
+pub struct ReadOptions {
+    /// A hint for prefix key to check bloom filter.
+    /// If the `prefix_hint` is not None, it should be included in
+    /// `key` or `key_range` in the read API.
+    prefix_hint: Option<Vec<u8>>,
```

Comment:
IMO, `prefix_hint` + `check_bloom_filter` it's a compromise of the previous implementation. It looks a little duplicated. 

We may encounter some confusing situations such as `prefix_hint` is not `None`, but  `check_bloom_filter` is false for `iter` interface. Its there a good way to unify the `point_get` and the`iter` with `ReadOption` ?

Reply:
Discussed offline. We can treat `check_bloom_filter` as a switch to turn on/off prefix bloom filter. We can change this when encountering with problems during implementation.